### DISCUSSION TITLE
How to Mass-Import Timesheet Entries?

### DISCUSSION THREAD

**Discussion by duyanhtruong** [kimai/kimai#4977] [Questions & Answers]
Hello everyone,

I setup Kimai with LDAP for self-hosting on my windows server.
I have hundreds of timesheet entries in JSON format ready to import through the Importer plugin. However, the limit per import is 1 MB or 1000 rows per file. If I do this manually it will take quite some time to get through hundreds of JSON files.

Is there a way to increase the file size limit for the Importer? Or automate mass-importing large amounts of JSON files? (I'm talking 600+ files, each with 1000 timesheet entries / or one big JSON file with around 60,000 entries).

There is nothing in the database - no project, no customers, users. I need projects/activities/customers/users to be automatically created upon the import, if the they don't already exist, like how the Importer plugin works.

Just for reference, the JSON timesheet entry format for the Importer plugin looks like this:
```
[
    {
        "Date": "2004-04-29",
        "From": "08:00",
        "To": "16:30",
        "Duration": 30600,
        "Rate": null,
        "User": "First Last",
        "Email": "first.last@company.com",
        "Customer": "Internal",
        "Project": "Project Name",
        "Activity": "Testing",
        "Description": "",
        "Exported": false,
        "Tags": "foo,bar,import",
        "HourlyRate": null,
        "FixedRate": 0,
        "InternalRate": null
    }
]
```
I don't have id numbers for projects/activtities/customers/users. I only have the info the above.

Apologies if this is already asked elsewhere, I went through many discussions but could not find a relevant one.
Your help is much appreciated!

**Comment by kevinpapst** [comment]
You can try to assign -1 as memory_limit for your PHP process and increase the row limit: https://github.com/kevinpapst/ImportBundle/blob/main/Importer/ImporterService.php#L23

**Reply by duyanhtruong** [reply]
Thank you for the speedy reply!

I set` memory_limit = -1` in php.ini and `MAX_ROWS = 1000000` in ImporterService.php
Reloaded cache restarted Apache.
Still didn't work: "The uploaded file was too large, Please try to upload a smaller file."

**Comment by duyanhtruong** [comment]
Ok, I found the config in another file: https://github.com/kevinpapst/ImportBundle/blob/main/Form/ImportForm.php
I changed `'maxSize' => '1024k'` to `'102400k'`

This worked to increase the file size limit from 1 MB to 2 MB. Still, anything above 2 MB it won't take.

**Reply by kevinpapst** [reply]
Oh thanks, I wasn't aware of that limit, need to make it configurable.
The limit is likely caused by PHP Max_upload_size core setting.

**Reply by duyanhtruong** [reply]
> Oh thanks, I wasn't aware of that limit, need to make it configurable. The limit is likely caused by PHP Max_upload_size core setting.

![image](https://github.com/user-attachments/assets/c4da7270-ef3e-4bec-b6ad-406dadd463ce)
In case it is of help in pinpointing the config location, this is the message I get when trying to upload a json file **slightly** above 2 MB. The message does not show up for file sizes significantly larger than 2 MB.

**Reply by kevinpapst** [reply]
Both PHP and your webserver can have an upload limit. Kimai does not have additional upload restrictions.

**Reply by duyanhtruong** [reply]
> The limit is likely caused by PHP Max_upload_size core setting.

You were right. I should've looked more closely at **php.ini**. I played around with these attributes:
```
memory_limit = -1
upload_max_filesize
post_max_size // should be higher than upload_max_filesize
max_execution_time // making sure to update this value to above 30 seconds to accommodate higher file sizes
```
The importer can import big files now. The largest I've tested is around 10 MB, but will continue to push for higher.

Thank you so much your help!

**Reply by kevinpapst** [reply]
Cool! Could you please share max rows and files sizes you were able to import?!

Would be great if you could test that with a real `memory_limit` of e.g. `256M`.
This needs to run in many environments and -1 is not a usual prod or shared-hosting setting.

**Reply by duyanhtruong** [reply]
Here's my test php.ini config:
```
memory_limit = 256M
max_execution_time = 30    // default value: 30
max_input_time = 60    // default value: 60
upload_max_filesize = 25M    // default value: 2M
post_max_size = 100M    // default value: 8M
```
The max JSON file size that this config seems to be comfortable with is around **6 MB (12000 entries in a single JSON file)**.

I then tested higher file size: 8 MB (15000 entries)
Set logging to debug level. Got this error output in var/log/prod.log when I tried the above config:
`[php] CRITICAL: Fatal Error: Maximum execution time of 30 seconds exceeded {"exception":"[object] (Symfony\\Component\\ErrorHandler\\Error\\FatalError(code: 0): Error: Maximum execution time of 30 seconds exceeded at Kimai\\vendor\\symfony\\http-foundation\\Response.php:425)"} []`

Going higher than 8 MB (e.g. 10 MB file) would result in the error:
`[php] CRITICAL: Fatal Error: Allowed memory size of 268435456 bytes exhausted (tried to allocate 83886088 bytes) {"exception":"[object] (Symfony\\Component\\ErrorHandler\\Error\\OutOfMemoryError(code: 0): Error: Allowed memory size of 268435456 bytes exhausted (tried to allocate 83886088 bytes) at Kimai\\vendor\\twig\\twig\\src\\Extension\\CoreExtension.php:1781)"} []`

Self-hosting Kimai on Windows server with 128 GB RAM

**Reply by kevinpapst** [reply]
Thank you! 
Sounds as if I could safely up the limit to 5k rows and 4MB upload size.
